### PR TITLE
Addition of RTS/CTS/RS485 capability

### PR DESCRIPTION
**Pull request by mubes**
These patches implement RTS/CTS and RS485 handling for UART by means of adding optional rts, cts, rs485_dir and rs485_invert options to the Python UART constructor.  Implementation is complete and compiles for all IMXRT range and was tested on IMXRT1021. For other CPUs it will compile but will throw an exception if any attempt is made to use the new features, until they're added in the underlying driver.

It also addresses a few bugs in the IMXRT UART handler.

### CODE REVIEW

**Comment by mubes** [comment]
RS485 now incorporated and tested.

File: ports/atmel-samd/common-hal/busio/UART.c [hunk 1]
```diff
@@ -52,16 +52,23 @@ static void usart_async_rxc_callback(const struct usart_async_descriptor *const
...
+    mp_float_t timeout, uint16_t receiver_buffer_size) {
+
     Sercom* sercom = NULL;
     uint8_t sercom_index = 255; // Unset index
     uint32_t rx_pinmux = 0;
     uint8_t rx_pad = 255; // Unset pad
     uint32_t tx_pinmux = 0;
     uint8_t tx_pad = 255; // Unset pad
 
+    if ((rts != mp_const_none) || (cts != mp_const_none) || (rs485_dir != mp_const_none) || (rs485_invert == true)) {
```
Comment (arturo182) [inline ports/atmel-samd/common-hal/busio/UART.c]: No need to do "== true"

File: ports/cxd56/common-hal/busio/UART.c [hunk 2]
```diff
@@ -53,11 +53,17 @@ STATIC busio_uart_dev_t busio_uart_dev[] = {
...
-    uint8_t bits, uart_parity_t parity, uint8_t stop, mp_float_t timeout,
-    uint16_t receiver_buffer_size) {
+    const mcu_pin_obj_t * tx, const mcu_pin_obj_t * rx,
+    const mcu_pin_obj_t * rts, const mcu_pin_obj_t * cts,
+    const mcu_pin_obj_t * rs485_dir, bool rs485_invert,
+    uint32_t baudrate, uint8_t bits, uart_parity_t parity, uint8_t stop,
+    mp_float_t timeout, uint16_t receiver_buffer_size) {
     struct termios tio;
 
+    if ((rts != mp_const_none) || (cts != mp_const_none) || (rs485_dir != mp_const_none) || (rs485_invert == true)) {
```
Comment (arturo182) [inline ports/cxd56/common-hal/busio/UART.c]: No need for  == true

File: ports/mimxrt10xx/common-hal/busio/UART.c [hunk 3]
```diff
@@ -111,12 +113,63 @@ void common_hal_busio_uart_construct(busio_uart_obj_t *self,
...
+    // Filter for sane settings for RS485
+    if (rs485_dir != mp_const_none) {
+      if ((rts != mp_const_none) || (cts != mp_const_none)) {
+        mp_raise_ValueError(translate("Cannot specify RTS or CTS in RS485 mode"));
+      }
+      // For IMXRT the RTS pin is used for RS485 direction
+      rts = rs485_dir;
+    }
+    else
+      {
```
Comment (arturo182) [inline ports/mimxrt10xx/common-hal/busio/UART.c]: Please fix the formatting here to } else {

File: ports/mimxrt10xx/common-hal/busio/UART.c [hunk 4]
```diff
@@ -111,12 +113,63 @@ void common_hal_busio_uart_construct(busio_uart_obj_t *self,
...
+        if (rs485_invert == true) {
```
Comment (arturo182) [inline ports/mimxrt10xx/common-hal/busio/UART.c]: No need to == true

File: ports/mimxrt10xx/common-hal/busio/UART.c [hunk 5]
```diff
@@ -111,12 +113,63 @@ void common_hal_busio_uart_construct(busio_uart_obj_t *self,
...
+        }
+      }
+
+    // Now check for RTS/CTS (or overloaded RS485 direction) pin(s)
+    const uint32_t rts_count = sizeof(mcu_uart_rts_list) / sizeof(mcu_periph_obj_t);
+    const uint32_t cts_count = sizeof(mcu_uart_cts_list) / sizeof(mcu_periph_obj_t);
+
+    if (rts != mp_const_none) {
+      for (uint32_t i=0; i < rts_count; ++i)
+        {
```
Comment (arturo182) [inline ports/mimxrt10xx/common-hal/busio/UART.c]: { in previous line (same in other places)

File: ports/mimxrt10xx/common-hal/busio/UART.c [hunk 6]
```diff
@@ -111,12 +113,63 @@ void common_hal_busio_uart_construct(busio_uart_obj_t *self,
...
+          if (mcu_uart_rts_list[i].bank_idx == self->rx_pin->bank_idx) {
+            if (mcu_uart_rts_list[i].pin == rts) {
+              self->rts_pin = &mcu_uart_rts_list[i];
+              break;
+            }
+          }
+        }
+      if (self->rts_pin == NULL)
```
Comment (arturo182) [inline ports/mimxrt10xx/common-hal/busio/UART.c]: Indent here is off

File: ports/mimxrt10xx/peripherals/mimxrt10xx/MIMXRT1011/periph.h [hunk 7]
```diff
@@ -36,6 +36,8 @@ extern const mcu_periph_obj_t mcu_spi_miso_list[4];
 
 extern const mcu_periph_obj_t mcu_uart_rx_list[9];
 extern const mcu_periph_obj_t mcu_uart_tx_list[9];
+const mcu_periph_obj_t mcu_uart_rts_list[4];
+const mcu_periph_obj_t mcu_uart_cts_list[4];
```
Comment (arturo182) [inline ports/mimxrt10xx/peripherals/mimxrt10xx/MIMXRT1011/periph.h]: extern like others

File: ports/mimxrt10xx/peripherals/mimxrt10xx/MIMXRT1021/periph.c [hunk 8]
```diff
@@ -109,53 +109,93 @@ LPUART_Type *mcu_uart_banks[] = { LPUART1, LPUART2, LPUART3, LPUART4, LPUART5, L
...
+    PERIPH_PIN(4, 2, 0, 1, &pin_GPIO_EMC_31),
+
+    PERIPH_PIN(5, 2, 0, 0, &pin_GPIO_EMC_37),
+
+    PERIPH_PIN(6, 2, 0, 0, &pin_GPIO_EMC_15),
+
+    PERIPH_PIN(7, 2, 0, 1, &pin_GPIO_SD_B0_03),
+
+    PERIPH_PIN(8, 2, 0, 0, &pin_GPIO_EMC_25),
+};
```
Comment (arturo182) [inline ports/mimxrt10xx/peripherals/mimxrt10xx/MIMXRT1021/periph.c]: Wow, is the SDK really missing all those ` kIOMUXC_*` defines?

File: ports/mimxrt10xx/peripherals/mimxrt10xx/MIMXRT1062/periph.h [hunk 9]
```diff
@@ -36,6 +36,8 @@ extern const mcu_periph_obj_t mcu_spi_miso_list[8];
 
 extern const mcu_periph_obj_t mcu_uart_rx_list[18];
 extern const mcu_periph_obj_t mcu_uart_tx_list[18];
+const mcu_periph_obj_t mcu_uart_rts_list[9];
+const mcu_periph_obj_t mcu_uart_cts_list[9];
```
Comment (arturo182) [inline ports/mimxrt10xx/peripherals/mimxrt10xx/MIMXRT1062/periph.h]: extern

**Comment by tannewt** [comment]
> I'm not as strong as @tannewt so I nitpick a bit about the formatting. Other than that, everything looks good! Were the pin fixes that you made in the 1011 and 1062 wrong originally?

Please be careful who you nitpick about formatting. It can deter some folks from contributing more and isn't worth the risk if we can automate that away anyway. (Code review comments on formatting also distract from deeper level review too.)

In this case, I think @mubes is happy to match whatever style we choose.

File: ports/mimxrt10xx/peripherals/mimxrt10xx/MIMXRT1021/periph.c (hunk 8 above)
Comment (mubes) [reply ports/mimxrt10xx/peripherals/mimxrt10xx/MIMXRT1021/periph.c]: This one I need to investigate....

Comment (mubes) [reply ports/mimxrt10xx/peripherals/mimxrt10xx/MIMXRT1021/periph.c]: These only exist where there's a daisy-chain register to configure...and for most of the RTS/CTS pins there isn't.

**Comment by mubes** [comment]
OK, that should be all the review comments addressed and merge issues addressed. Hopefully ready for shipping.

**Comment by arturo182** [comment]
> (Code review comments on formatting also distract from deeper level review too.)

Yes, I've seen that happen quite a few times in some projects. My way for dealing with that is to make 2 passes at the code, first one is the obvious formatting issues and then I take a second pass looking only at the functionality of the code and comment on that. 

Yes I understand we don't want to scare people off with comments about formatting but we also want to have nice code :D

**Comment by arturo182** [comment]
Looks good! @mubes please run `make translate` and add the .po files to the PR.

**Comment by mubes** [comment]
> > (Code review comments on formatting also distract from deeper level review too.)
> 
> Yes, I've seen that happen quite a few times in some projects. My way for dealing with that is to make 2 passes at the code, first one is the obvious formatting issues and then I take a second pass looking only at the functionality of the code and comment on that.
> 
> Yes I understand we don't want to scare people off with comments about formatting but we also want to have nice code :D

That was the idea behind putting the files up a few days back before they were fully cooked. I have no problem making coding style changes (e.g. ==true tests) because consistency there really helps code understanding but format and layout things should be dealt with by a tool, even if the results are less than perfect layout-wise. Personally, I don't want to be wasting highly qualified and skilled people on making sure their brackets are on the right line. That _is_ important for clear and consistent understanding of the code, but it can, and should, be done by machine. We need to get a formatter up and integrated ASAP...then the comments only need to address the substantive code style issues.

**Pull request merged**